File: task3/reconciliation.js
#!/usr/bin/env node

/**
 * Payment Reconciliation Script
 * Reconciles expected payments with actual payment records to identify discrepancies.
 */

const fs = require('fs');
const path = require('path');

class PaymentReconciler {
    constructor(expectedFile, actualFile, fuzzyThreshold = 0.7) {
        this.expectedFile = expectedFile;
        this.actualFile = actualFile;
        this.fuzzyThreshold = fuzzyThreshold;
        this.expectedPayments = [];
        this.actualPayments = [];
        this.ambiguityRules = {
            'M. Chen': 'Michael Chen',
            'E. Davis': 'Emma Davis',
            'Unknown Donor': 'UNMATCHED',
        };
    }

    // Parse CSV content
    parseCSV(content) {
        const lines = content.trim().split('\n');
        const headers = lines[0].split(',').map(h => h.trim());
        const records = [];

        for (let i = 1; i < lines.length; i++) {
            const values = lines[i].split(',').map(v => v.trim());
            const record = {};
            headers.forEach((header, idx) => {
                record[header] = values[idx];
            });
            records.push(record);
        }
        return records;
    }

    // Load CSV files
    loadFiles() {
        const expectedContent = fs.readFileSync(this.expectedFile, 'utf-8');
        const actualContent = fs.readFileSync(this.actualFile, 'utf-8');

        this.expectedPayments = this.parseCSV(expectedContent);
        this.actualPayments = this.parseCSV(actualContent);
    }

    // Calculate string similarity
    fuzzyMatch(name1, name2) {
        const n1 = name1.toLowerCase().trim();
        const n2 = name2.toLowerCase().trim();

        if (n1 === n2) return 1;

        let matches = 0;
        const minLen = Math.min(n1.length, n2.length);

        for (let i = 0; i < minLen; i++) {
            if (n1[i] === n2[i]) matches++;
        }

        return matches / Math.max(n1.length, n2.length);
    }

    // Resolve ambiguous names
    resolveName(payerName) {
        if (this.ambiguityRules[payerName]) {
            return this.ambiguityRules[payerName];
        }

        // Fuzzy match against expected names
        for (const expected of this.expectedPayments) {
            const expectedName = expected.Name;
            const similarity = this.fuzzyMatch(payerName, expectedName);
            if (similarity >= this.fuzzyThreshold) {
                return expectedName;
            }
        }

        return payerName;
    }

    // Perform reconciliation
    reconcile() {
        this.loadFiles();

        const results = {
            matched: [],
            unmatchedExpected: [],
            unmatchedActual: [],
            duplicates: [],
            amountMismatches: [],
            extraPayments: [],
            followUpList: []
        };

        // Create dictionaries for tracking
        const expectedDict = {};
        for (const exp of this.expectedPayments) {
            const name = exp.Name;
            const amount = parseFloat(exp.Amount);
            if (!expectedDict[name]) {
                expectedDict[name] = { count: 0, total: 0, entries: [] };
            }
            expectedDict[name].total += amount;
            expectedDict[name].count += 1;
            expectedDict[name].entries.push(exp);
        }

        const actualDict = {};
        for (const act of this.actualPayments) {
            const resolvedName = this.resolveName(act['Payer Name']);
            if (!actualDict[resolvedName]) {
                actualDict[resolvedName] = [];
            }
            actualDict[resolvedName].push(act);
        }

        // Match payments
        const matchedActual = new Set();

        for (const name in expectedDict) {
            const expectedAmount = expectedDict[name].total;

            if (actualDict[name]) {
                const actualRecords = actualDict[name];
                const actualTotal = actualRecords.reduce((sum, r) => sum + parseFloat(r.Amount), 0);

                // Check for duplicates
                if (actualRecords.length > 1 && actualTotal > expectedAmount) {
                    for (let i = 0; i < actualRecords.length; i++) {
                        const record = actualRecords[i];
                        results.duplicates.push({
                            name: name,
                            amount: record.Amount,
                            date: record.Date,
                            notes: record['Reference/Notes'],
                            index: i
                        });
                    }
                }

                // Check amount mismatch
                if (Math.abs(actualTotal - expectedAmount) > 0.01) {
                    results.amountMismatches.push({
                        name: name,
                        expected: expectedAmount,
                        actual: actualTotal,
                        difference: actualTotal - expectedAmount
                    });
                } else {
                    results.matched.push({
                        name: name,
                        amount: expectedAmount,
                        records: actualRecords.length
                    });
                }

                // Mark as matched
                for (const record of actualRecords) {
                    matchedActual.add(JSON.stringify(record));
                }
            } else {
                results.unmatchedExpected.push({
                    name: name,
                    amount: expectedAmount
                });
            }
        }

        // Find unmatched actual payments
        for (const name in actualDict) {
            const records = actualDict[name];
            for (const record of records) {
                if (!matchedActual.has(JSON.stringify(record))) {
                    let found = false;
                    for (const expName in expectedDict) {
                        if (this.fuzzyMatch(name, expName) >= this.fuzzyThreshold) {
                            found = true;
                            break;
                        }
                    }

                    if (!found || !expectedDict[name]) {
                        results.extraPayments.push({
                            name: name,
                            amount: record.Amount,
                            date: record.Date,
                            method: record['Payment Method'],
                            notes: record['Reference/Notes']
                        });
                    }
                }
            }
        }

        // Generate follow-up list
        for (const item of results.unmatchedExpected) {
            results.followUpList.push({
                type: 'MISSING_PAYMENT',
                name: item.name,
                amount: item.amount,
                action: `Contact ${item.name} for payment of $${parseFloat(item.amount).toFixed(2)}`
            });
        }

        for (const item of results.amountMismatches) {
            results.followUpList.push({
                type: 'AMOUNT_MISMATCH',
                name: item.name,
                amount: item.expected,
                action: `Verify payment for ${item.name}: Expected $${item.expected.toFixed(2)}, Received $${item.actual.toFixed(2)} (Diff: $${item.difference.toFixed(2)})`
            });
        }

        for (const item of results.duplicates) {
            results.followUpList.push({
                type: 'DUPLICATE',
                name: item.name,
                amount: item.amount,
                action: `Review duplicate entry for ${item.name} on ${item.date} ($${item.amount}) - ${item.notes}`
            });
        }

        return results;
    }

    // Calculate totals
    calculateTotals() {
        const totalExpected = this.expectedPayments.reduce((sum, p) => sum + parseFloat(p.Amount), 0);
        const totalActual = this.actualPayments.reduce((sum, p) => sum + parseFloat(p.Amount), 0);
        return [totalExpected, totalActual];
    }

    // Print formatted report
    printReport(results) {
        const [totalExpected, totalActual] = this.calculateTotals();
        const now = new Date().toISOString().replace('T', ' ').substring(0, 19);

        console.log('\n' + '='.repeat(80));
        console.log('PAYMENT RECONCILIATION REPORT');
        console.log('='.repeat(80));
        console.log(`Generated: ${now}`);
        console.log();

        // Summary
        console.log('SUMMARY');
        console.log('-'.repeat(80));
        console.log(`Total Expected: $${totalExpected.toFixed(2).padStart(12)}`);
        console.log(`Total Recorded: $${totalActual.toFixed(2).padStart(12)}`);
        const diff = totalActual - totalExpected;
        const sign = diff >= 0 ? '+' : '';
        console.log(`Discrepancy:    $${sign}${diff.toFixed(2).padStart(11)}`);
        console.log();

        // Matched Payments
        console.log(`MATCHED PAYMENTS (${results.matched.length})`);
        console.log('-'.repeat(80));
        for (const item of results.matched) {
            const name = item.name.padEnd(40, '.');
            const amount = `$${parseFloat(item.amount).toFixed(2)}`.padStart(10);
            console.log(`  ✓ ${name} ${amount}`);
        }
        console.log();

        // Unmatched Expected
        if (results.unmatchedExpected.length > 0) {
            console.log(`MISSING PAYMENTS (${results.unmatchedExpected.length})`);
            console.log('-'.repeat(80));
            for (const item of results.unmatchedExpected) {
                const name = item.name.padEnd(40, '.');
                const amount = `$${parseFloat(item.amount).toFixed(2)}`.padStart(10);
                console.log(`  ✗ ${name} ${amount}`);
            }
            console.log();
        }

        // Duplicates
        if (results.duplicates.length > 0) {
            console.log(`DUPLICATE ENTRIES (${results.duplicates.length})`);
            console.log('-'.repeat(80));
            for (const item of results.duplicates) {
                const name = item.name.padEnd(30, '.');
                console.log(`  ⚠ ${name} $${item.amount} on ${item.date}`);
                console.log(`     Notes: ${item.notes}`);
            }
            console.log();
        }

        // Amount Mismatches
        if (results.amountMismatches.length > 0) {
            console.log(`AMOUNT MISMATCHES (${results.amountMismatches.length})`);
            console.log('-'.repeat(80));
            for (const item of results.amountMismatches) {
                console.log(`  ${item.name.padEnd(30)}`);
                const expected = `$${item.expected.toFixed(2)}`.padStart(12);
                const actual = `$${item.actual.toFixed(2)}`.padStart(12);
                const diff = `$${item.difference.toFixed(2)}`.padStart(12);
                console.log(`    Expected: ${expected}  |  Received: ${actual}  |  Diff: ${diff}`);
            }
            console.log();
        }

        // Extra/Unmatched Actual
        if (results.extraPayments.length > 0) {
            console.log(`EXTRA/UNIDENTIFIED PAYMENTS (${results.extraPayments.length})`);
            console.log('-'.repeat(80));
            for (const item of results.extraPayments) {
                const name = item.name.padEnd(30, '.');
                console.log(`  ${name} $${item.amount} (${item.method})`);
                console.log(`    Date: ${item.date}  |  Notes: ${item.notes}`);
            }
            console.log();
        }

        // Follow-up List
        console.log(`FOLLOW-UP REQUIRED (${results.followUpList.length})`);
        console.log('-'.repeat(80));
        for (let i = 0; i < results.followUpList.length; i++) {
            const item = results.followUpList[i];
            const typeStr = item.type.padEnd(15, '.');
            console.log(`  ${i + 1}. [${typeStr}] ${item.action}`);
        }
        console.log();

        console.log('='.repeat(80));
        console.log('END OF REPORT');
        console.log('='.repeat(80));
    }
}

// Main execution
if (require.main === module) {
    const reconciler = new PaymentReconciler(
        'expected_payments.csv',
        'payment_records.csv',
        0.7
    );

    try {
        const results = reconciler.reconcile();
        reconciler.printReport(results);
    } catch (error) {
        console.error('Error during reconciliation:', error.message);
        process.exit(1);
    }
}

module.exports = PaymentReconciler;

File: task3/reconciliation.py
#!/usr/bin/env python3
"""
Payment Reconciliation Script
Reconciles expected payments with actual payment records to identify discrepancies.
"""

import csv
from difflib import SequenceMatcher
from datetime import datetime
from collections import defaultdict

class PaymentReconciler:
    def __init__(self, expected_file, actual_file, fuzzy_threshold=0.7):
        self.expected_file = expected_file
        self.actual_file = actual_file
        self.fuzzy_threshold = fuzzy_threshold
        self.expected_payments = []
        self.actual_payments = []
        self.ambiguity_rules = {
            "M. Chen": "Michael Chen",
            "E. Davis": "Emma Davis",
            "Unknown Donor": "UNMATCHED",
        }

    def load_files(self):
        """Load expected and actual payment records from CSV files."""
        with open(self.expected_file, 'r') as f:
            reader = csv.DictReader(f)
            self.expected_payments = list(reader)

        with open(self.actual_file, 'r') as f:
            reader = csv.DictReader(f)
            self.actual_payments = list(reader)

    def fuzzy_match(self, name1, name2):
        """Calculate similarity between two names."""
        name1_lower = name1.lower().strip()
        name2_lower = name2.lower().strip()
        return SequenceMatcher(None, name1_lower, name2_lower).ratio()

    def resolve_name(self, payer_name):
        """Resolve ambiguous payer names using predefined rules."""
        if payer_name in self.ambiguity_rules:
            return self.ambiguity_rules[payer_name]

        # Fuzzy match against expected names
        for expected_payment in self.expected_payments:
            expected_name = expected_payment['Name']
            similarity = self.fuzzy_match(payer_name, expected_name)
            if similarity >= self.fuzzy_threshold:
                return expected_name

        return payer_name

    def reconcile(self):
        """Perform reconciliation and identify discrepancies."""
        self.load_files()

        results = {
            'matched': [],
            'unmatched_expected': [],
            'unmatched_actual': [],
            'duplicates': [],
            'amount_mismatches': [],
            'extra_payments': [],
            'follow_up_list': []
        }

        # Create dictionaries for tracking
        expected_dict = {}
        for exp in self.expected_payments:
            name = exp['Name']
            amount = float(exp['Amount'])
            if name not in expected_dict:
                expected_dict[name] = {'count': 0, 'total': 0, 'entries': []}
            expected_dict[name]['total'] += amount
            expected_dict[name]['count'] += 1
            expected_dict[name]['entries'].append(exp)

        actual_dict = defaultdict(list)
        for act in self.actual_payments:
            resolved_name = self.resolve_name(act['Payer Name'])
            actual_dict[resolved_name].append(act)

        # Match payments
        matched_actual = set()

        for name in expected_dict:
            expected_amount = expected_dict[name]['total']

            if name in actual_dict:
                actual_records = actual_dict[name]
                actual_total = sum(float(r['Amount']) for r in actual_records)

                # Check for duplicates
                if len(actual_records) > 1 and actual_total > expected_amount:
                    for i, record in enumerate(actual_records):
                        results['duplicates'].append({
                            'name': name,
                            'amount': record['Amount'],
                            'date': record['Date'],
                            'notes': record['Reference/Notes'],
                            'index': i
                        })

                # Check amount mismatch
                if abs(actual_total - expected_amount) > 0.01:  # Account for rounding
                    results['amount_mismatches'].append({
                        'name': name,
                        'expected': expected_amount,
                        'actual': actual_total,
                        'difference': actual_total - expected_amount
                    })
                else:
                    results['matched'].append({
                        'name': name,
                        'amount': expected_amount,
                        'records': len(actual_records)
                    })

                # Mark as matched
                for record in actual_records:
                    matched_actual.add(tuple(record.items()))
            else:
                results['unmatched_expected'].append({
                    'name': name,
                    'amount': expected_amount
                })

        # Find unmatched actual payments
        for name, records in actual_dict.items():
            for record in records:
                if tuple(record.items()) not in matched_actual:
                    # Check if it's an unmatched expected payment that was found
                    found = False
                    for exp_name in expected_dict:
                        if self.fuzzy_match(name, exp_name) >= self.fuzzy_threshold:
                            found = True
                            break

                    if not found or name not in expected_dict:
                        results['extra_payments'].append({
                            'name': name,
                            'amount': record['Amount'],
                            'date': record['Date'],
                            'method': record['Payment Method'],
                            'notes': record['Reference/Notes']
                        })

        # Generate follow-up list
        for item in results['unmatched_expected']:
            results['follow_up_list'].append({
                'type': 'MISSING_PAYMENT',
                'name': item['name'],
                'amount': item['amount'],
                'action': f"Contact {item['name']} for payment of ${item['amount']}"
            })

        for item in results['amount_mismatches']:
            results['follow_up_list'].append({
                'type': 'AMOUNT_MISMATCH',
                'name': item['name'],
                'amount': item['expected'],
                'action': f"Verify payment for {item['name']}: Expected ${item['expected']}, Received ${item['actual']} (Diff: ${item['difference']:.2f})"
            })

        for item in results['duplicates']:
            results['follow_up_list'].append({
                'type': 'DUPLICATE',
                'name': item['name'],
                'amount': item['amount'],
                'action': f"Review duplicate entry for {item['name']} on {item['date']} (${item['amount']}) - {item['notes']}"
            })

        return results

    def calculate_totals(self):
        """Calculate total expected and actual amounts."""
        total_expected = sum(float(p['Amount']) for p in self.expected_payments)
        total_actual = sum(float(p['Amount']) for p in self.actual_payments)
        return total_expected, total_actual

    def print_report(self, results):
        """Print a formatted reconciliation report."""
        total_expected, total_actual = self.calculate_totals()

        print("\n" + "="*80)
        print("PAYMENT RECONCILIATION REPORT")
        print("="*80)
        print(f"Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
        print()

        # Summary
        print("SUMMARY")
        print("-" * 80)
        print(f"Total Expected: ${total_expected:,.2f}")
        print(f"Total Recorded: ${total_actual:,.2f}")
        print(f"Discrepancy:    ${total_actual - total_expected:+,.2f}")
        print()

        # Matched Payments
        print("MATCHED PAYMENTS ({})".format(len(results['matched'])))
        print("-" * 80)
        for item in results['matched']:
            print(f"  ✓ {item['name']:.<40} ${item['amount']:>10,.2f}")
        print()

        # Unmatched Expected
        if results['unmatched_expected']:
            print("MISSING PAYMENTS ({})".format(len(results['unmatched_expected'])))
            print("-" * 80)
            for item in results['unmatched_expected']:
                print(f"  ✗ {item['name']:.<40} ${item['amount']:>10,.2f}")
            print()

        # Duplicates
        if results['duplicates']:
            print("DUPLICATE ENTRIES ({})".format(len(results['duplicates'])))
            print("-" * 80)
            for item in results['duplicates']:
                print(f"  ⚠ {item['name']:.<30} ${item['amount']:>10} on {item['date']}")
                print(f"     Notes: {item['notes']}")
            print()

        # Amount Mismatches
        if results['amount_mismatches']:
            print("AMOUNT MISMATCHES ({})".format(len(results['amount_mismatches'])))
            print("-" * 80)
            for item in results['amount_mismatches']:
                print(f"  {item['name']:.<30}")
                print(f"    Expected: ${item['expected']:>10,.2f}  |  Received: ${item['actual']:>10,.2f}  |  Diff: ${item['difference']:+10,.2f}")
            print()

        # Extra/Unmatched Actual
        if results['extra_payments']:
            print("EXTRA/UNIDENTIFIED PAYMENTS ({})".format(len(results['extra_payments'])))
            print("-" * 80)
            for item in results['extra_payments']:
                print(f"  {item['name']:.<30} ${item['amount']:>10} ({item['method']})")
                print(f"    Date: {item['date']}  |  Notes: {item['notes']}")
            print()

        # Follow-up List
        print("FOLLOW-UP REQUIRED ({})".format(len(results['follow_up_list'])))
        print("-" * 80)
        for i, item in enumerate(results['follow_up_list'], 1):
            print(f"  {i}. [{item['type']:.<15}] {item['action']}")
        print()

        print("="*80)
        print("END OF REPORT")
        print("="*80)


if __name__ == "__main__":
    reconciler = PaymentReconciler(
        expected_file="expected_payments.csv",
        actual_file="payment_records.csv",
        fuzzy_threshold=0.7
    )

    results = reconciler.reconcile()
    reconciler.print_report(results)
